Release/src/http/client/http_client_curl.cpp

```diff
@@ -0,0 +1,2379 @@
+                            done = true;
+                        }
+
+                        // Let the caller process the result
+                        break;
+                    }
+
+                    // Move on to the next byte
+                    n++;
+                }
+
+                offset = n;
+                length = l;
+                return buffer_size ? done : (!m_bytes_remaining && !m_chunk_size && !m_chunk_delim);
+            }
+
+        private:
+            size_t m_bytes_remaining; // the number of bytes remaining in the chunk we're currently processing
+            bool m_chunk_size;        // if true, we're processing a chunk size or its trailing delimiter
+            bool m_chunk_delim;       // if true, we're processing a delimiter between a chunk and the next chunk's size
+            bool m_expect_linefeed; // if true, we're processing a delimiter, and we've already seen its carriage return
+            bool m_ignore;          // if true, we're processing a chunk extension or trailer, which we don't support
+            bool m_trailer;         // if true, we're processing (and ignoring) a trailer field; m_ignore is also true
+        };
+
+        std::vector<uint8_t> m_buffer; // we read data from the stream into this before compressing
+        uint8_t* m_acquired; // we use this in place of m_buffer if the stream has directly-accessible data available
+        size_t m_bytes_read; // we most recently read this many bytes, which may be less than m_buffer.size()
+        size_t m_bytes_processed; // we've compressed this many bytes of m_bytes_read so far
+        bool m_needs_flush;   // we've read and compressed all bytes, but the compressor still has compressed bytes to
+                              // give us
+        bool m_started;       // we've sent at least some number of bytes to m_decompressor
+        bool m_done;          // we've read, compressed, and consumed all bytes
+        bool m_chunked;       // if true, we need to decode and decompress a transfer-encoded message
+        size_t m_chunk_bytes; // un-decompressed bytes remaining in the most-recently-obtained data from m_chunk
+        std::unique_ptr<_chunk_helper> m_chunk;
+    } m_compression_state;
+
+    void cleanup()
+    {
+        if (m_compression_state.m_acquired != nullptr)
+        {
+            // We may still hold a piece of the buffer if we encountered an exception; release it here
+            if (m_decompressor)
+            {
+                _get_writebuffer().commit(0);
+            }
+            else
+            {
+                _get_readbuffer().release(m_compression_state.m_acquired, m_compression_state.m_bytes_processed);
+            }
+            m_compression_state.m_acquired = nullptr;
+        }
+
+        if (m_request_handle != nullptr)
+        {
+            WinHttpCloseHandle(m_request_handle);
+        }
+    }
+
+    void install_custom_cn_check(const utility::string_t& customHost)
```

Comment:
I don't think you need any of this CN handling for curl -- AFAIK it already follows the "Host: header for CN negotiation" thing including for TLS-SNI. But I'm not 100% sure.

Reply:
CURL does the header validation internally. It doesn't expose an API to force override.

The following test passes: outside_tests:server_hostname_mismatch